src/sources/apache_metrics/mod.rs

```diff
@@ -0,0 +1,403 @@
+    internal_events::{
+        ApacheMetricsErrorResponse, ApacheMetricsEventReceived, ApacheMetricsHttpError,
+        ApacheMetricsParseError, ApacheMetricsRequestCompleted,
+    },
+    shutdown::ShutdownSignal,
+    Event, Pipeline,
+};
+use chrono::Utc;
+use futures::{
+    compat::{Future01CompatExt, Sink01CompatExt},
+    future, stream, FutureExt, StreamExt, TryFutureExt,
+};
+use futures01::Sink;
+use hyper::{Body, Client, Request};
+use hyper_openssl::HttpsConnector;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::BTreeMap;
+use std::time::{Duration, Instant};
+
+mod parser;
+
+pub use parser::ParseError;
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+struct ApacheMetricsConfig {
+    endpoints: Vec<String>,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+    #[serde(default = "default_namespace")]
+    namespace: String,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+pub fn default_namespace() -> String {
+    "apache".to_string()
+}
+
+inventory::submit! {
+    SinkDescription::new_without_default::<ApacheMetricsConfig>("apache_metrics")
+}
+
+#[typetag::serde(name = "apache_metrics")]
+impl crate::config::SourceConfig for ApacheMetricsConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        let urls = self
+            .endpoints
+            .iter()
+            .map(|endpoint| endpoint.parse::<http::Uri>())
+            .collect::<Result<Vec<_>, _>>()
+            .context(super::UriParseError)?;
+
+        Ok(apache_metrics(
+            urls,
+            self.scrape_interval_secs,
+            self.namespace.clone(),
+            shutdown,
+            out,
+        ))
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "apache_metrics"
+    }
+}
+
+fn apache_metrics(
+    urls: Vec<http::Uri>,
+    interval: u64,
+    namespace: String,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> super::Source {
+    let out = out
+        .sink_map_err(|e| error!("error sending metric: {:?}", e))
+        .sink_compat();
+    let task = tokio::time::interval(Duration::from_secs(interval))
+        .take_until(shutdown.compat())
+        .map(move |_| stream::iter(urls.clone()))
+        .flatten()
+        .map(move |url| {
+            let https = HttpsConnector::new().expect("TLS initialization failed");
+            let client = Client::builder().build(https);
+
+            let request = Request::get(&url)
+                .body(Body::empty())
+                .expect("error creating request");
+
+            let mut tags: BTreeMap<String, String> = BTreeMap::new();
+            tags.insert("endpoint".into(), url.to_string());
```

Comment:
From [rfc/2020-08-21-3092-apache-metrics-source.md](https://github.com/timberio/vector/blob/3d11ab33f6a3ae3de6e2b1a73f97ca23879720a8/rfcs/2020-08-21-3092-apache-metrics-source.md):
> `endpoint` the full endpoint (sans any basic auth credentials)

Should we remove credentials here?

Reply:
Yes we should :smile: Thanks for catching that!

Reply:
I started to fix this, but will have to pick it back up later. It wasn't as straightforward as I thought it might be.

Reply:
https://github.com/timberio/vector/pull/4410